contracts/Faucet.sol

```diff
@@ -9,113 +9,105 @@ import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
 contract Faucet is Ownable {
     using SafeERC20 for IERC20;
 
+    uint256 public constant TIMEOUT_LIMIT = 30;
+    uint256 public constant MAX_WITHDRAWAL_AMOUNT = 5 ether;
     address public contractRegistry;
-    address public bctAddress;
-    address public nctAddress;
-    mapping(address => uint256) private lastWithdrawalTimes;
+    mapping(address => bool) public isPoolEligible;
+    mapping(address => uint256) private _lastWithdrawalTimes;
     event Deposited(address erc20Addr, uint256 amount);
     event Withdrawn(address account, address erc20Addr, uint256 amount);
 
-    constructor(
-        address _contractRegistry,
-        address _bctAddress,
-        address _nctAddress
-    ) {
-        contractRegistry = _contractRegistry;
-        bctAddress = _bctAddress;
-        nctAddress = _nctAddress;
+    constructor(address contractRegistry_, address[] memory poolAddresses) {
+        contractRegistry = contractRegistry_;
+
+        uint256 poolAddressesCount = poolAddresses.length;
+        for (uint256 i = 0; i < poolAddressesCount; i++) {
+            setPoolEligible(poolAddresses[i], true);
+        }
     }
 
     /// @notice change the TCO2 contracts registry
-    /// @param _address the new contract registry address
+    /// @param contractRegistry_ the new contract registry address
     function setToucanContractRegistry(
-        address _address
+        address contractRegistry_
+    ) public virtual onlyOwner {
+        contractRegistry = contractRegistry_;
+    }
+
+    /// @notice allows the owner to set the eligibility of a pool
+    /// @param poolAddress the address of the pool
+    /// @param isPoolEligible_ eligibility flag
+    function setPoolEligible(
+        address poolAddress,
+        bool isPoolEligible_
     ) public virtual onlyOwner {
-        contractRegistry = _address;
+        isPoolEligible[poolAddress] = isPoolEligible_;
```

Comment:
Apart from this mapping (and potential instead of), we likely should be storing pools in an array that the frontend can pull to check which tokens the user holds to deposit or can withdraw without the need to extend Subgraph for this type of functionality.

Reply:
the mapping was to optimize the eligibility check during deposits and the addition of a new pool, to avoid a O(n) check every time. 

Rather than keeping 2 data structures containing the same information though, I would rather go with only the array at this point and swallow the additional gas cost, although it's a) testnet so not really a big deal, b) the number of pools won't likely grow to really problematic levels.

What do you say>

Reply:
This was my thinking too. I guess an alternative to making any further changes here would be for the faucet UI to fetch all pool tokens held by the faucet contract via the Subgraph, then we don't need to do pool token discovery with the use of this array.

Reply:
The alternative is to keep everything as is and hardcode the list on the frontend. Sounds dirty, but if the list doesn't change very frequently (like less than once per quarter), it would save gas and RPC calls all over the place.

Reply:
I don't think there is any need to hardcode anything to the frontend. There are two options that avoid hardcoding:
1. expose arrays of supported tokens in the contract. Then the UI can do one call to fetch the supported tokens, then a second multicall to fetch balances for the user and the faucet contract
2. use Subgraph in the faucet UI (if not used already) to fetch token balances for both the faucet and UI

Reply:
Just to summarize the subgraph flow with the proposals you made:

Pool Whitelisting:
- owner add an eligible Pool to the faucet, whose method shall emit a "AddPool" event
- subgraph catches the event and whitelists the pool

TCO2 whitelisting:
- factory deploys a new token and emits `TokenCreated`
- subgraph listens to this event and whitelists it

Deposit:
- UI fetches the list of whitelisted tokens and shows them and their respective balances to the holder
- token holder selects one of them transfers tco2/pool tokens to faucet, simply via `transferFrom`
- subgraph catches the `Transfer` event and keeps a running count for the given pool/tco2 updated

Withdraw flow:
- UI fetches the list of whitelisted tokens and shows them and their respective balances to the user
- user issues a `withdraw` request
- subgraph catches the `Transfer` event and decreases the count accordingly 


Is that in line with your idea?

Reply:
This would indeed imply also removing `Deposit` and `Withdraw`